### DISCUSSION TITLE
MIDI Thru Options

### DISCUSSION THREAD

**Discussion by spiffo-smith** [RomanKubiak/ctrlr#542] [General]
Does anyone have a comprehensive knowledge of what all the Ctrlr MIDI Thru Options do?

The only one I have ever found a use for is:
**Plugin host -> Output device**

When the Panel is a VST this allows MIDI Notes from a DAW MIDI Track to pass THRU the Panel whilst still allowing the Panel function simultaneously as an Editor, sending SysEx or whatever. Very useful for a Synth Editor!

Just wondering if I am missing out on anything, does anyone else use any of the other options and why?

**Comment by thethirdapple** [comment]
While certainly not comprehensive, my limited experience for what its worth:

Using  Plogue Bidule software, I have found some interesting routing options with simply a "blank panel" functioning as a patchbay and could be really useful when combined with some Lua scripts.

Found this thread on the the OG Forum referencing the midi options including the "Comparator":
https://ctrlr.org/forums/topic/midi-settings-ui-needed/

BTW - It is due to your videos Part 1-4 which lead me to finally begin building a CTRLR panel prior to knowing much about LUA scripting.

Peace

![midi_thru_patchbay](https://user-images.githubusercontent.com/125695472/222488214-110f78e6-bc78-4b9f-b599-107c2dd2937c.png)

**Reply by spiffo-smith** [reply]
First Off thanks for the compliment on the YouTube Videos!

I found a second possible use:
**Input Device -> Output Device** and Input Device Channel set to **All Channels**

This will allow a MIDI Keyboard setup as a **Controller Device** to pass MIDI Note On/Off on any MIDI Channel straight THRU to the MIDI Output Device. Useful for passing all Note ON/OFF on any Channel to a MultiTimbral Synth!

Although the logic of setting the Input Device Channel to **All Channels** is lost on me? If you want to pass any MIDI Channel from the Controller to the Output Device, you would think one of those ought to be set to **All Channels** not the Input Device which should have nothing to do with it? That has to be a Bug?

**Comment by damiensellier** [comment]
Hi guys, all the different modes are in the following script : 
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/UIComponents/CtrlrMIDI/CtrlrMIDISettingsRouting.cpp#L42

IN Device  -> OUT Device : "All messages that are received on the MIDI Input device of CTRLR will be copied to the selected output MIDI Device of CTRLR"

CTRL Device -> OUT Device : "All messages that are received on the MIDI Controller device of CTRLR will be copied to the selected output MIDI Device of CTRLR"

HOST -> OUT Device : "All messages received by CTRLR from the plugin HOST, will be copied to the MIDI Output device of CTRLR"

HOST -> HOST : "All messages received  by CTRLR from the plugin HOST, will be copied back to the plugin HOST"


**HOST -> OUT Device** as far as I understand, can be helpful if you are on Windows and don't have a multi instance MIDI soundcard, it would prevents duplicating the message with using a midi loopback software and a MIDI splitter program to avoid those "cannot open", "midi device not available" or "midi device already in use" (something like that). Ctrlr doesn't act as a second app but a parrallel one. It can be great for listening and displaying/analysing the incoming midi messages from a Host MIDI track for exemple.
You create a midi track in the host, select the MIDI input device from there, enable the channel MIDI IN, the Midi track messages will go to the CTRLR plugin. When CTRLR plugin will receive the midi messages from the host midi track, it will route those messages to the device output and channel of your choice you can set in the CTRLR midi menu.

If it doesn't work like that I'm totally confused :)

NOTE: from what I can see in the comparator and THRU function, only the MIDI messages are passed, it means SYSEX won't be transfered.

**Reply by spiffo-smith** [reply]
Thanks for the link to the C Plus Plus. Unfortunately for me it doesn't really clear anything up!

I have a MIDI Keyboard connected via USB MIDI (one MIDI Device) and a Roland MIDI Interface (In and Out) connected to a JV-2080.
I have the MIDI Keyboard set as the **Controller Device** and the Roland MIDI Interface set as the **Input Device** and **Output Device**

If I want to pass all MIDI Notes (on any Channel) from the Keyboard through to the JV-2080 I have to Tick **IN Device -> OUT Device**
Ticking **CTRL Device -> OUT Device** does nothing! If that is supposed to make logical sense then I must have very illogical brain!

5.3.201 and 5.6.23 appear to behave the same way.

I'll check the **HOST -> OUT Device** function like you said that sounds like it could be useful!

**Reply by spiffo-smith** [reply]
OK, like you said damiensellier:

**Third Use:**
In a VST Plugin inside a DAW, with no **Input, Controller or Output MIDI Device** selected at all in the Panel, if you Tick either:
**HOST -> OUT** Device or **HOST -> HOST** then the incoming MIDI Notes can be sniffed in the **MIDI Monitor!**

Perhaps SysEx Messages could be sniffed as well but my DAW blocks them as do most I believe.

**Reply by damiensellier** [reply]
Yeah exactly, you should be able to get the MIDI messages going from the DAW to CTRLR and reroute them to the CTRLR MIDI Out port without any changes or a MIDI Channel modification. The way the code is written in the Ctrlr core is more like a midi transform and routing tool. but it first filters for midi compliant messages only so no sysex should go thru unfortunately. Then it has this channelize function to override the source midi channel to the one you set in the menu.

I never used it but I can imagine some use for that though.

**Reply by dervur** [reply]
I can't seem to have the CTRLR device in my options...
![image](https://github.com/user-attachments/assets/e84878ea-5161-46bb-847a-94b568a7fcef)

Does anybody know how else could I change the channel of the input by following the output channel? E.g., I play C3 with Maschine set to Ch.1, the panel out is set to Ch.3, I want that C3 note to be sent to Ch.3 and not Ch.1 by the panel. 
The MIDI thru channelize doesn't do anything or at least it seems it doesn't, or I'm missing something. Alternatively is it possible to use the MIDI input device to control the uiMidiKeyboard? This would solve what I want to do too I think as the uiMidiKeyboard will send output messages instead (which follow the channel output selection).

Also taking the chance to thank spiffo for the videos that got me into Ctrlr!
